File: Game-1-modular/systems/attack_effects.py
"""
Attack Effects System — Tag-driven visual feedback for combat.

Replaces the old blue/red line system with prominent, tag-colored effects:
- Slash arcs (melee weapons)
- Impact bursts (on hit)
- Blocked indicators
- Area-of-effect rings

All colors are driven by damage type tags (fire, ice, etc.).
Effects are rendered by the Renderer and fade out over time.
"""

import math
import time
from typing import List, Tuple, Optional, Dict, Any
from dataclasses import dataclass, field
from enum import Enum


class AttackEffectType(Enum):
    """Types of attack effects"""
    SLASH_ARC = "slash_arc"       # Sweeping arc effect (replaces old LINE)
    IMPACT_BURST = "impact_burst" # Radial burst at hit point
    BLOCKED = "blocked"           # Attack blocked indicator
    AREA = "area"                 # Circle effect for AoE attacks
    THRUST = "thrust"             # Forward thrust line (spears, daggers)


class AttackSourceType(Enum):
    """Source of the attack for fallback color"""
    PLAYER = "player"
    TURRET = "turret"
    ENEMY = "enemy"
    ENVIRONMENT = "environment"


# Tag -> color mapping for element-driven visuals
_ELEMENT_COLORS = {
    "physical": (220, 220, 240),
    "slashing": (230, 230, 250),
    "crushing": (200, 200, 220),
    "piercing": (210, 220, 240),
    "fire": (255, 120, 30),
    "ice": (100, 200, 255),
    "frost": (100, 200, 255),
    "lightning": (255, 255, 80),
    "poison": (100, 255, 80),
    "arcane": (180, 80, 255),
    "shadow": (130, 80, 180),
    "holy": (255, 255, 180),
}

# Source type -> fallback color (only used when no tags)
_SOURCE_FALLBACK_COLORS = {
    AttackSourceType.PLAYER: (150, 200, 255),
    AttackSourceType.TURRET: (0, 220, 255),
    AttackSourceType.ENEMY: (255, 80, 60),
    AttackSourceType.ENVIRONMENT: (255, 150, 50),
}


def color_from_tags(tags: List[str],
                    source_type: AttackSourceType = AttackSourceType.PLAYER) -> Tuple[int, int, int]:
    """Get effect color from tags, falling back to source-based color."""
    for tag in tags:
        if tag in _ELEMENT_COLORS:
            return _ELEMENT_COLORS[tag]
    return _SOURCE_FALLBACK_COLORS.get(source_type, (220, 220, 240))


@dataclass
class AttackEffect:
    """A visual effect to render. Tag-driven colors."""
    effect_type: AttackEffectType
    source_type: AttackSourceType
    start_pos: Tuple[float, float]     # World coordinates (attacker)
    end_pos: Tuple[float, float]       # World coordinates (target/endpoint)
    start_time: float
    duration: float = 1.05             # Seconds (3x longer for visible animation)
    blocked: bool = False
    damage: float = 0.0
    tags: List[str] = field(default_factory=list)
    facing_angle: float = 0.0          # Degrees, for arc direction
    arc_degrees: float = 55.0          # Width of slash arc (default moderate)
    radius: float = 1.5               # Reach of the effect in tiles

    @property
    def age(self) -> float:
        return time.time() - self.start_time

    @property
    def alpha(self) -> float:
        age = self.age
        if age >= self.duration:
            return 0.0
        # Quick bright start, smooth fade
        fade_start = self.duration * 0.5
        if age < fade_start:
            return 1.0
        return 1.0 - ((age - fade_start) / (self.duration - fade_start))

    @property
    def is_expired(self) -> bool:
        return self.age >= self.duration

    def get_color(self) -> Tuple[int, int, int, int]:
        """Get RGBA color from tags."""
        if self.blocked:
            return (255, 200, 0, int(self.alpha * 255))
        base = color_from_tags(self.tags, self.source_type)
        return (*base, int(self.alpha * 255))

    def get_line_width(self) -> int:
        """Scale width with damage for impact effects."""
        base = 3
        if self.damage > 50:
            base = 5
        elif self.damage > 20:
            base = 4
        return max(2, int(base * max(0.4, self.alpha)))


class AttackEffectsManager:
    """Manages tag-driven attack visual effects."""

    def __init__(self):
        self.effects: List[AttackEffect] = []
        self.max_effects = 100

    def add_attack_effect(self,
                          source_pos: Tuple[float, float],
                          target_pos: Tuple[float, float],
                          source_type: AttackSourceType,
                          damage: float = 0.0,
                          blocked: bool = False,
                          tags: Optional[List[str]] = None,
                          duration: float = 1.5,
                          facing_angle: float = 0.0,
                          arc_degrees: float = 55.0,
                          radius: float = 1.5):
        """Add a prominent slash arc or thrust effect.

        Automatically picks SLASH_ARC for melee or THRUST for piercing/spear.
        """
        tags = tags or ['physical']

        # Determine effect type from tags
        is_thrust = any(t in tags for t in ('piercing', 'spear', 'thrust', 'dagger'))
        effect_type = AttackEffectType.THRUST if is_thrust else AttackEffectType.SLASH_ARC

        if blocked:
            effect_type = AttackEffectType.BLOCKED

        effect = AttackEffect(
            effect_type=effect_type,
            source_type=source_type,
            start_pos=source_pos,
            end_pos=target_pos,
            start_time=time.time(),
            duration=duration,
            blocked=blocked,
            damage=damage,
            tags=tags,
            facing_angle=facing_angle,
            arc_degrees=arc_degrees,
            radius=radius,
        )
        self._add_effect(effect)

    def add_attack_line(self,
                        source_pos: Tuple[float, float],
                        target_pos: Tuple[float, float],
                        source_type: AttackSourceType,
                        damage: float = 0.0,
                        blocked: bool = False,
                        tags: Optional[List[str]] = None,
                        duration: float = 1.05):
        """Backward-compatible: converts old line calls to slash arcs."""
        dx = target_pos[0] - source_pos[0]
        dy = target_pos[1] - source_pos[1]
        angle = math.degrees(math.atan2(dy, dx))
        dist = math.sqrt(dx * dx + dy * dy)
        self.add_attack_effect(
            source_pos, target_pos, source_type, damage, blocked,
            tags, duration, facing_angle=angle, radius=max(1.0, dist))

    def add_impact_burst(self,
                         position: Tuple[float, float],
                         source_type: AttackSourceType,
                         damage: float = 0.0,
                         tags: Optional[List[str]] = None,
                         duration: float = 1.35):
        """Add a radial impact burst at hit location. Prominent and bright."""
        tags = tags or ['physical']
        effect = AttackEffect(
            effect_type=AttackEffectType.IMPACT_BURST,
            source_type=source_type,
            start_pos=position,
            end_pos=position,
            start_time=time.time(),
            duration=duration,
            damage=damage,
            tags=tags,
            radius=max(0.5, damage / 30.0),  # Scale with damage
        )
        self._add_effect(effect)

    def add_blocked_indicator(self,
                              position: Tuple[float, float],
                              source_type: AttackSourceType,
                              duration: float = 1.5):
        """Add blocked attack indicator."""
        effect = AttackEffect(
            effect_type=AttackEffectType.BLOCKED,
            source_type=source_type,
            start_pos=position,
            end_pos=position,
            start_time=time.time(),
            duration=duration,
            blocked=True,
        )
        self._add_effect(effect)

    def add_area_effect(self,
                        center_pos: Tuple[float, float],
                        radius: float,
                        source_type: AttackSourceType,
                        tags: Optional[List[str]] = None,
                        duration: float = 1.8):
        """Add AoE ring effect."""
        tags = tags or ['physical']
        effect = AttackEffect(
            effect_type=AttackEffectType.AREA,
            source_type=source_type,
            start_pos=center_pos,
            end_pos=(center_pos[0] + radius, center_pos[1]),
            start_time=time.time(),
            duration=duration,
            tags=tags,
            radius=radius,
        )
        self._add_effect(effect)

    def _add_effect(self, effect: AttackEffect):
        self.effects.append(effect)
        if len(self.effects) > self.max_effects:
            self.effects = [e for e in self.effects if not e.is_expired]

    def update(self):
        self.effects = [e for e in self.effects if not e.is_expired]

    def get_active_effects(self) -> List[AttackEffect]:
        return [e for e in self.effects if not e.is_expired]

    def clear(self):
        self.effects.clear()


# Global singleton
_attack_effects_manager: Optional[AttackEffectsManager] = None


def get_attack_effects_manager() -> AttackEffectsManager:
    """Get the global attack effects manager instance."""
    global _attack_effects_manager
    if _attack_effects_manager is None:
        _attack_effects_manager = AttackEffectsManager()
    return _attack_effects_manager

File: Unity/Assets/Scripts/Game1.Systems/Combat/AttackEffects.cs
// ============================================================================
// Game1.Systems.Combat.AttackEffects
// Migrated from: systems/attack_effects.py (233 lines)
// Migration phase: 4
// Date: 2026-02-13
// ============================================================================
//
// Visual feedback data for combat: attack lines, blocked indicators,
// hit particles, and AoE area effects.
//
// This is LOGIC ONLY -- no rendering. The renderer reads the effect data
// and draws it. Effects fade over time and are automatically pruned.
//
// Color mapping:
//   Player  = Blue   (50, 150, 255)
//   Turret  = Cyan   (0, 220, 255)
//   Enemy   = Red    (255, 50, 50)
//   Blocked = Yellow (255, 200, 0)
//   Environment = Orange (255, 150, 50)
//
// NO MonoBehaviour, NO UnityEngine. Pure C#.
// ============================================================================

using System;
using System.Collections.Generic;
using Game1.Data.Models;

namespace Game1.Systems.Combat
{
    /// <summary>
    /// Types of attack visual effects.
    /// Python: AttackEffectType enum.
    /// </summary>
    public enum AttackEffectType
    {
        /// <summary>A line from attacker to target.</summary>
        Line,

        /// <summary>Attack blocked indicator (X mark at collision point).</summary>
        Blocked,

        /// <summary>Small particles at hit location.</summary>
        HitParticle,

        /// <summary>Circle effect for AoE attacks.</summary>
        Area
    }

    /// <summary>
    /// Source of the attack, determines color.
    /// Python: AttackSourceType enum.
    /// </summary>
    public enum AttackSourceType
    {
        /// <summary>Player attacks are blue (50, 150, 255).</summary>
        Player,

        /// <summary>Turret attacks are cyan (0, 220, 255).</summary>
        Turret,

        /// <summary>Enemy attacks are red (255, 50, 50).</summary>
        Enemy,

        /// <summary>Environment/other attacks are orange (255, 150, 50).</summary>
        Environment
    }

    /// <summary>
    /// RGBA color tuple (0-255 per channel).
    /// Pure C# replacement for Pygame/Unity Color.
    /// </summary>
    public readonly struct ColorRgba
    {
        public byte R { get; }
        public byte G { get; }
        public byte B { get; }
        public byte A { get; }

        public ColorRgba(byte r, byte g, byte b, byte a)
        {
            R = r;
            G = g;
            B = b;
            A = a;
        }
    }

    /// <summary>
    /// A visual attack effect with position, timing, and appearance data.
    /// Rendered by the UI layer; this class is logic-only.
    ///
    /// Python: @dataclass AttackEffect (lines 34-99)
    /// </summary>
    public class AttackEffect
    {
        /// <summary>Type of visual effect.</summary>
        public AttackEffectType EffectType { get; set; }

        /// <summary>Source type for color determination.</summary>
        public AttackSourceType SourceType { get; set; }

        /// <summary>Start position in world coordinates (attacker or center).</summary>
        public GamePosition StartPos { get; set; }

        /// <summary>End position in world coordinates (target, or same as start for non-lines).</summary>
        public GamePosition EndPos { get; set; }

        /// <summary>Game time when the effect was created.</summary>
        public float StartTime { get; set; }

        /// <summary>
        /// Duration in seconds before full fade out.
        /// Python: duration: float = 0.3
        /// </summary>
        public float Duration { get; set; } = 0.3f;

        /// <summary>Whether this attack was blocked.</summary>
        public bool Blocked { get; set; } = false;

        /// <summary>Damage amount (for scaling effect intensity).</summary>
        public float Damage { get; set; } = 0f;

        /// <summary>Attack tags for special effects (e.g., fire, ice).</summary>
        public List<string> Tags { get; set; } = new();

        /// <summary>
        /// Get the age of this effect in seconds relative to the given time.
        /// Python: @property age -> time.time() - self.start_time
        /// </summary>
        /// <param name="currentTime">Current game time.</param>
        /// <returns>Age in seconds.</returns>
        public float GetAge(float currentTime)
        {
            return currentTime - StartTime;
        }

        /// <summary>
        /// Get the alpha value (0.0-1.0) for fading.
        /// Quick fade-in, slow fade-out starting at 70% of duration.
        ///
        /// Python: @property alpha (lines 54-62)
        ///   fade_start = self.duration * 0.7
        ///   if age < fade_start: return 1.0
        ///   return 1.0 - ((age - fade_start) / (self.duration - fade_start))
        /// </summary>
        /// <param name="currentTime">Current game time.</param>
        /// <returns>Alpha 0.0 (invisible) to 1.0 (fully visible).</returns>
        public float GetAlpha(float currentTime)
        {
            float age = GetAge(currentTime);

            if (age >= Duration)
                return 0f;

            // Quick fade in, slow fade out
            float fadeStart = Duration * 0.7f;
            if (age < fadeStart)
                return 1.0f;

            return 1.0f - ((age - fadeStart) / (Duration - fadeStart));
        }

        /// <summary>
        /// Check if this effect should be removed.
        /// Python: @property is_expired -> self.age >= self.duration
        /// </summary>
        /// <param name="currentTime">Current game time.</param>
        /// <returns>True if effect is expired.</returns>
        public bool IsExpired(float currentTime)
        {
            return GetAge(currentTime) >= Duration;
        }

        /// <summary>
        /// Get RGBA color based on source type and blocked state.
        ///
        /// Python: get_color() (lines 69-88)
        ///   Blocked = yellow (255, 200, 0)
        ///   Player  = blue   (50, 150, 255)
        ///   Turret  = cyan   (0, 220, 255)
        ///   Enemy   = red    (255, 50, 50)
        ///   Other   = orange (255, 150, 50)
        /// </summary>
        /// <param name="currentTime">Current game time (for alpha calculation).</param>
        /// <returns>RGBA color with current alpha.</returns>
        public ColorRgba GetColor(float currentTime)
        {
            byte alpha = (byte)(GetAlpha(currentTime) * 255);

            if (Blocked)
            {
                return new ColorRgba(255, 200, 0, alpha);
            }

            return SourceType switch
            {
                AttackSourceType.Player => new ColorRgba(50, 150, 255, alpha),
                AttackSourceType.Turret => new ColorRgba(0, 220, 255, alpha),
                AttackSourceType.Enemy => new ColorRgba(255, 50, 50, alpha),
                AttackSourceType.Environment => new ColorRgba(255, 150, 50, alpha),
                _ => new ColorRgba(255, 150, 50, alpha)
            };
        }

        /// <summary>
        /// Get line width based on damage amount (scaled by alpha).
        ///
        /// Python: get_line_width() (lines 90-99)
        ///   damage > 50 -> base 4
        ///   damage > 20 -> base 3
        ///   else -> base 2
        ///   return max(1, int(base_width * self.alpha))
        /// </summary>
        /// <param name="currentTime">Current game time (for alpha).</param>
        /// <returns>Line width in pixels (1-4).</returns>
        public int GetLineWidth(float currentTime)
        {
            int baseWidth = 2;
            if (Damage > 50f)
                baseWidth = 4;
            else if (Damage > 20f)
                baseWidth = 3;

            return Math.Max(1, (int)(baseWidth * GetAlpha(currentTime)));
        }
    }

    // GamePosition is defined in Game1.Data.Models — use that canonical definition.

    /// <summary>
    /// Manages all active attack visual effects.
    /// Provides methods to add effects and query/prune active ones.
    ///
    /// Python: AttackEffectsManager (lines 102-233)
    ///
    /// Usage:
    ///   manager.AddAttackLine(sourcePos, targetPos, sourceType, damage);
    ///   manager.Update(currentTime);
    ///   var effects = manager.GetActiveEffects(currentTime);
    /// </summary>
    public class AttackEffectsManager
    {
        /// <summary>All tracked effects (including potentially expired ones).</summary>
        private readonly List<AttackEffect> _effects = new();

        /// <summary>
        /// Maximum number of effects before pruning.
        /// Python: self.max_effects = 100
        /// </summary>
        public int MaxEffects { get; set; } = 100;

        // ====================================================================
        // Add Effects
        // ====================================================================

        /// <summary>
        /// Add an attack line effect from source to target.
        /// Python: add_attack_line()
        /// </summary>
        /// <param name="sourcePos">World position of attacker.</param>
        /// <param name="targetPos">World position of target.</param>
        /// <param name="sourceType">Source type for color.</param>
        /// <param name="currentTime">Current game time.</param>
        /// <param name="damage">Damage amount for scaling.</param>
        /// <param name="blocked">Whether attack was blocked.</param>
        /// <param name="tags">Attack tags for special effects.</param>
        /// <param name="duration">Effect duration (default 0.3s).</param>
        public void AddAttackLine(GamePosition sourcePos, GamePosition targetPos,
                                  AttackSourceType sourceType, float currentTime,
                                  float damage = 0f, bool blocked = false,
                                  List<string> tags = null, float duration = 0.3f)
        {
            var effect = new AttackEffect
            {
                EffectType = AttackEffectType.Line,
                SourceType = sourceType,
                StartPos = sourcePos,
                EndPos = targetPos,
                StartTime = currentTime,
                Duration = duration,
                Blocked = blocked,
                Damage = damage,
                Tags = tags ?? new List<string>()
            };

            _addEffect(effect);
        }

        /// <summary>
        /// Add a blocked attack indicator at a position.
        /// Python: add_blocked_indicator()
        /// </summary>
        /// <param name="position">World position where attack was blocked.</param>
        /// <param name="sourceType">Source type for context.</param>
        /// <param name="currentTime">Current game time.</param>
        /// <param name="duration">Indicator duration (default 0.5s).</param>
        public void AddBlockedIndicator(GamePosition position, AttackSourceType sourceType,
                                        float currentTime, float duration = 0.5f)
        {
            var effect = new AttackEffect
            {
                EffectType = AttackEffectType.Blocked,
                SourceType = sourceType,
                StartPos = position,
                EndPos = position,
                StartTime = currentTime,
                Duration = duration,
                Blocked = true
            };

            _addEffect(effect);
        }

        /// <summary>
        /// Add an area effect indicator (for AoE attacks).
        /// Python: add_area_effect()
        ///
        /// Note: EndPos.X encodes radius (EndPos = center + (radius, 0)).
        /// This matches the Python hack: end_pos=(center_pos[0] + radius, center_pos[1]).
        /// </summary>
        /// <param name="centerPos">Center of AoE in world coordinates.</param>
        /// <param name="radius">Radius of effect in tiles.</param>
        /// <param name="sourceType">Source type for color.</param>
        /// <param name="currentTime">Current game time.</param>
        /// <param name="duration">Effect duration (default 0.4s).</param>
        public void AddAreaEffect(GamePosition centerPos, float radius,
                                  AttackSourceType sourceType, float currentTime,
                                  float duration = 0.4f)
        {
            var effect = new AttackEffect
            {
                EffectType = AttackEffectType.Area,
                SourceType = sourceType,
                StartPos = centerPos,
                EndPos = new GamePosition(centerPos.X + radius, centerPos.Y, centerPos.Z),
                StartTime = currentTime,
                Duration = duration,
                Damage = 0f,
                Tags = new List<string> { "circle" }
            };

            _addEffect(effect);
        }

        /// <summary>
        /// Get the radius from an area effect.
        /// Computed from EndPos.X - StartPos.X.
        /// </summary>
        public static float GetAreaRadius(AttackEffect effect)
        {
            return effect.EndPos.X - effect.StartPos.X;
        }

        // ====================================================================
        // Update / Query
        // ====================================================================

        /// <summary>
        /// Remove expired effects.
        /// Python: update() — self.effects = [e for e in self.effects if not e.is_expired]
        /// </summary>
        /// <param name="currentTime">Current game time.</param>
        public void Update(float currentTime)
        {
            _effects.RemoveAll(e => e.IsExpired(currentTime));
        }

        /// <summary>
        /// Get all currently active (non-expired) effects.
        /// Python: get_active_effects()
        /// </summary>
        /// <param name="currentTime">Current game time.</param>
        /// <returns>List of active effects for rendering.</returns>
        public List<AttackEffect> GetActiveEffects(float currentTime)
        {
            var active = new List<AttackEffect>();
            foreach (var effect in _effects)
            {
                if (!effect.IsExpired(currentTime))
                    active.Add(effect);
            }
            return active;
        }

        /// <summary>
        /// Get the number of active effects.
        /// </summary>
        public int ActiveCount(float currentTime)
        {
            int count = 0;
            foreach (var effect in _effects)
            {
                if (!effect.IsExpired(currentTime))
                    count++;
            }
            return count;
        }

        /// <summary>
        /// Clear all effects.
        /// Python: clear()
        /// </summary>
        public void Clear()
        {
            _effects.Clear();
        }

        // ====================================================================
        // Private
        // ====================================================================

        /// <summary>
        /// Add an effect, pruning old ones if over the max limit.
        /// Python: _add_effect()
        /// </summary>
        private void _addEffect(AttackEffect effect)
        {
            _effects.Add(effect);

            // Prune expired if over limit
            // Python: if len(self.effects) > self.max_effects:
            //             self.effects = [e for e in self.effects if not e.is_expired]
            if (_effects.Count > MaxEffects)
            {
                // Use effect's StartTime + Duration to check expiry without a currentTime param
                // This is a conservative prune -- some effects might still be alive
                float pruneTime = effect.StartTime;
                _effects.RemoveAll(e => e.IsExpired(pruneTime));
            }
        }
    }
}
